### PR TITLE
fix(s3exporter/config-validation): endpoint should contain region and S3 bucket

### PR DESCRIPTION
**Pull request by dabcoder**
**Description:**

Discussion: https://github.com/open-telemetry/opentelemetry-collector-contrib/discussions/32766.

**Link to tracking Issue:** #32774

**Testing:** added test cases to validate config when endpoint is provided.

**Documentation:** updated README.

Configuration logic:

- An endpoint is provided, and Region still needs to be provided (as per the AWS SDK Go config, even though the region can be included in the endpoint itself). The S3 bucket name is optional. 
- An endpoint is not provided, thus region **and** S3 bucket name should be provided (return an error for each if one of those 2 isn't)

Tasks:

- [x] Add entry to changelog
- [x] Edit the docs

### CODE REVIEW

File: exporter/awss3exporter/config.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ func (c *Config) Validate() error {
 	if c.S3Uploader.Region == "" {
 		errs = multierr.Append(errs, errors.New("region is required"))
 	}
-	if c.S3Uploader.S3Bucket == "" {
+	if c.S3Uploader.S3Bucket == "" && c.S3Uploader.Endpoint == "" {
```
Comment (atoulme) [inline exporter/awss3exporter/config.go:53]: should it have its own error?

Comment (atoulme) [reply exporter/awss3exporter/config.go:53]: wait, should it be and or or?

File: exporter/awss3exporter/config_test.go [hunk 2]
```diff
@@ -108,13 +107,25 @@ func TestConfig_Validate(t *testing.T) {
 		config      *Config
 		errExpected error
 	}{
+		// Valid config with endpoint provided
 		{
-			name: "valid",
+			// endpoint should contain region and bucket name
```
Comment (atoulme) [inline exporter/awss3exporter/config_test.go:112]:
```suggestion
			// endpoint will be built from region and bucket name
```

Comment (atoulme) [inline exporter/awss3exporter/config_test.go:112]:
```suggestion
			// endpoint overrides region and bucket name.

File: exporter/awss3exporter/config_test.go [hunk 3]
```diff
@@ -124,32 +135,42 @@ func TestConfig_Validate(t *testing.T) {
 			config: func() *Config {
 				c := createDefaultConfig().(*Config)
 				c.S3Uploader.Region = ""
+				c.S3Uploader.S3Bucket = ""
+				c.S3Uploader.Endpoint = ""
 				return c
 			}(),
-			errExpected: multierr.Append(errors.New("region is required"),
-				errors.New("bucket is required")),
+			errExpected: errors.New("endpoint and region should be provided, or bucket amd region"),
```
Comment (atoulme) [inline exporter/awss3exporter/config_test.go:142]: this change is unexpected

Comment (dabcoder) [reply exporter/awss3exporter/config_test.go:142]: Good point, will revert.

Comment (dabcoder) [reply exporter/awss3exporter/config_test.go]: [5a23120](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32815/commits/5a23120ccee51aaa96f025c667d2d2bb7c8498fd)

File: exporter/awss3exporter/config_test.go (hunk 2 above)
Comment (dabcoder) [reply exporter/awss3exporter/config_test.go]: [7b60ae7](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32815/commits/7b60ae7e027de11b78c43d49db528949adb9ecba)

File: exporter/awss3exporter/config_test.go [hunk 4]
```diff
@@ -108,13 +107,25 @@ func TestConfig_Validate(t *testing.T) {
...
+			config: func() *Config {
+				c := createDefaultConfig().(*Config)
+				c.S3Uploader.Endpoint = "http://example.com"
+				c.S3Uploader.Region = "foo"
+				return c
+			}(),
+			errExpected: nil,
+		},
+		{
+			name: "valid with S3Bucket and region",
```
[the remark this answers is not in the document]
Comment (dabcoder) [reply exporter/awss3exporter/config_test.go:125]: [5a23120](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32815/commits/5a23120ccee51aaa96f025c667d2d2bb7c8498fd)

**Comment by dabcoder** [comment]
Actually I am not sure what protocol `endpoint` should use. I am seeing `unsupported protocol scheme \"s3\""}` errors in logs when using an endpoint starting with `s3://`, so we may also need to clarify that this support `http(s)` only?

**Comment by dabcoder** [comment]
cc @atoulme @pdelewski.

**Comment by atoulme** [comment]
> Actually I am not sure what protocol `endpoint` should use. I am seeing `unsupported protocol scheme \"s3\""}` errors in logs when using an endpoint starting with `s3://`, so we may also need to clarify that this support `http(s)` only?

Yes, I guess so.

**Review by bogdandrutu — APPROVED** [review]

**Comment by dabcoder** [comment]
@bogdandrutu anything else I should do in order to merge this?

**Comment by dabcoder** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 14 days.

Making sure that this remains live for the time being.

**Comment by dabcoder** [comment]
@open-telemetry/collector-contrib-maintainer if anyone has a chance to take another look at this PR before it gets marked as stale again, that'd be great, thanks.

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
@dabcoder PTAL at the failing tests

**Comment by dabcoder** [comment]
Thanks @dmitryax. They should be fixed now, I ran the tests locally and they're passing.

**Pull request merged**